# Stroke Dataset Cleansing

Data Cleansing for `healthcare-dataset-stroke-data.csv` using Python libraries.

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [40]:
df = pd.read_csv('healthcare-dataset-stroke-data.csv')
df

id  gender   age  hypertension  heart_disease ever_married  \
0      9046    Male  67.0             0              1          Yes   
1     51676  Female  61.0             0              0          Yes   
2     31112    Male  80.0             0              1          Yes   
3     60182  Female  49.0             0              0          Yes   
4      1665  Female  79.0             1              0          Yes   
...     ...     ...   ...           ...            ...          ...   
5105  18234  Female  80.0             1              0          Yes   
5106  44873  Female  81.0             0              0          Yes   
5107  19723  Female  35.0             0              0          Yes   
5108  37544    Male  51.0             0              0          Yes   
5109  44679  Female  44.0             0              0          Yes   

          work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0           Private          Urban             228.69  36.6  formerly smoked   
1     Self-employed          Rural             202.21   NaN     never smoked   
2           Private          Rural             105.92  32.5     never smoked   
3           Private          Urban             171.23  34.4           smokes   
4     Self-employed          Rural             174.12  24.0     never smoked   
...             ...            ...                ...   ...              ...   
5105        Private          Urban              83.75   NaN     never smoked   
5106  Self-employed          Urban             125.20  40.0     never smoked   
5107  Self-employed          Rural              82.99  30.6     never smoked   
5108        Private          Rural             166.29  25.6  formerly smoked   
5109       Govt_job          Urban              85.28  26.2          Unknown   

      stroke  
0          1  
1          1  
2          1  
3          1  
4          1  
...      ...  
5105       0  
5106       0  
5107       0  
5108       0  
5109       0  

[5110 rows x 12 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4909 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 4909 non-null   int64  
 1   gender             4909 non-null   object 
 2   age                4909 non-null   float64
 3   hypertension       4909 non-null   int64  
 4   heart_disease      4909 non-null   int64  
 5   ever_married       4909 non-null   object 
 6   work_type          4909 non-null   object 
 7   Residence_type     4909 non-null   object 
 8   avg_glucose_level  4909 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     4909 non-null   object 
 11  stroke             4909 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 498.6+ KB


# Cleaning Data

In [41]:
df.duplicated().sum()

0

In [42]:
df.isna().sum()

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

There are about 201 NaN data which we will then delete.

## Missing Value

In [7]:
nan_data = df[df.isna().any(axis=1)]
nan_data

id  gender   age  hypertension  heart_disease ever_married  \
1     51676  Female  61.0             0              0          Yes   
8     27419  Female  59.0             0              0          Yes   
13     8213    Male  78.0             0              1          Yes   
19    25226    Male  57.0             0              1           No   
27    61843    Male  58.0             0              0          Yes   
...     ...     ...   ...           ...            ...          ...   
5039  42007    Male  41.0             0              0           No   
5048  28788    Male  40.0             0              0          Yes   
5093  32235  Female  45.0             1              0          Yes   
5099   7293    Male  40.0             0              0          Yes   
5105  18234  Female  80.0             1              0          Yes   

          work_type Residence_type  avg_glucose_level  bmi   smoking_status  \
1     Self-employed          Rural             202.21  NaN     never smoked   
8           Private          Rural              76.15  NaN          Unknown   
13          Private          Urban             219.84  NaN          Unknown   
19         Govt_job          Urban             217.08  NaN          Unknown   
27          Private          Rural             189.84  NaN          Unknown   
...             ...            ...                ...  ...              ...   
5039        Private          Rural              70.15  NaN  formerly smoked   
5048        Private          Urban             191.15  NaN           smokes   
5093       Govt_job          Rural              95.02  NaN           smokes   
5099        Private          Rural              83.94  NaN           smokes   
5105        Private          Urban              83.75  NaN     never smoked   

      stroke  
1          1  
8          1  
13         1  
19         1  
27         1  
...      ...  
5039       0  
5048       0  
5093       0  
5099       0  
5105       0  

[201 rows x 12 columns]

In [19]:
df = df.dropna()

## Data Transformation

In [11]:
from sklearn.preprocessing import LabelEncoder
encode = LabelEncoder()

In [25]:
# Encoding

df['gender'] = encode.fit_transform(df['gender'].values)
df['ever_married'] = encode.fit_transform(df['ever_married'].values)
df['Residence_type'] = encode.fit_transform(df['Residence_type'].values)
df['smoking_status'] = encode.fit_transform(df['smoking_status'].values)
df['work_type'] = encode.fit_transform(df['work_type'].values)

In [22]:
# Drop unnecessary data
# There is unusual 'Others' data in the gender column so we need to drop it

df[df['gender']>1]
df.drop(df.index[df['gender']>1],inplace=True)

In [27]:
# Drop unused data

df = df.drop(['id'],axis=1)

In [28]:
df

gender   age  hypertension  heart_disease  ever_married  work_type  \
0          1  67.0             0              1             1          2   
2          1  80.0             0              1             1          2   
3          0  49.0             0              0             1          2   
4          0  79.0             1              0             1          3   
5          1  81.0             0              0             1          2   
...      ...   ...           ...            ...           ...        ...   
5104       0  13.0             0              0             0          4   
5106       0  81.0             0              0             1          3   
5107       0  35.0             0              0             1          3   
5108       1  51.0             0              0             1          2   
5109       0  44.0             0              0             1          0   

      Residence_type  avg_glucose_level   bmi  smoking_status  stroke  
0                  1             228.69  36.6               1       1  
2                  0             105.92  32.5               2       1  
3                  1             171.23  34.4               3       1  
4                  0             174.12  24.0               2       1  
5                  1             186.21  29.0               1       1  
...              ...                ...   ...             ...     ...  
5104               0             103.08  18.6               0       0  
5106               1             125.20  40.0               2       0  
5107               0              82.99  30.6               2       0  
5108               0             166.29  25.6               1       0  
5109               1              85.28  26.2               0       0  

[4908 rows x 11 columns]

In [29]:
# Scaling

from sklearn.preprocessing import MinMaxScaler
scale = MinMaxScaler()

In [32]:
df_col = [col for col in df.columns]

In [38]:
scaled = scale.fit_transform(df[df_col])

df_scaled = pd.DataFrame(scaled,columns=df_col)

In [39]:
df_scaled

gender       age  hypertension  heart_disease  ever_married  work_type  \
0        1.0  0.816895           0.0            1.0           1.0       0.50   
1        1.0  0.975586           0.0            1.0           1.0       0.50   
2        0.0  0.597168           0.0            0.0           1.0       0.50   
3        0.0  0.963379           1.0            0.0           1.0       0.75   
4        1.0  0.987793           0.0            0.0           1.0       0.50   
...      ...       ...           ...            ...           ...        ...   
4903     0.0  0.157715           0.0            0.0           0.0       1.00   
4904     0.0  0.987793           0.0            0.0           1.0       0.75   
4905     0.0  0.426270           0.0            0.0           1.0       0.75   
4906     1.0  0.621582           0.0            0.0           1.0       0.50   
4907     0.0  0.536133           0.0            0.0           1.0       0.00   

      Residence_type  avg_glucose_level       bmi  smoking_status  stroke  
0                1.0           0.801265  0.301260        0.333333     1.0  
1                0.0           0.234512  0.254296        0.666667     1.0  
2                1.0           0.536008  0.276060        1.000000     1.0  
3                0.0           0.549349  0.156930        0.666667     1.0  
4                1.0           0.605161  0.214204        0.333333     1.0  
...              ...                ...       ...             ...     ...  
4903             0.0           0.221402  0.095074        0.000000     0.0  
4904             1.0           0.323516  0.340206        0.666667     0.0  
4905             0.0           0.128658  0.232532        0.666667     0.0  
4906             0.0           0.513203  0.175258        0.333333     0.0  
4907             1.0           0.139230  0.182131        0.000000     0.0  

[4908 rows x 11 columns]